New TPOT features

kuratsak opened the pull request:
## What does this PR do?
Implements 3 cool new features:
1. A new optional argument that saves the best pipeline so far after each generation. Run TPOT -> use pipelines without stopping it's optimization process.
2. TPOT now saves the score that mattered for every optimized pipeline it outputs
3. personal/manual scoring function now supported in the main driver too.


## Where should the reviewer start?
Relatively small PR, should be straightforward.. just look at the full diff :dancer: 


## How should this PR be tested?
1. create a local_pipeline_folder
2. run TPOT and add -psp="./local_pipeline_folder/" 
3. after 2-3 generations TPOT will output at least one pipeline into the folder


## Any background context you want to provide?
Extremely useful features we used in our research projects!


## What are the relevant issues?

## Screenshots (if appropriate)



## Questions:

- Do the docs need to be updated? Nope
- Does this PR add new (Python) dependencies? Nope

kuratsak commented:
Hi Approvers!

This is my second attempt after splitting the pull request into a bug fix one that was approved already.

I am finishing a job and starting another, So I won't be able to fix code until at least July probably..
I have thoroughly tested and used these features in my research in the current job, I really hope we can approve this.

I am available until sunday for any fixes to be made in this pull request, otherwise it will have to wait and I really hope we don't have to.

weixuanfu2016 commented on tpot/driver.py:
```diff
@@ -404,6 +429,28 @@ def _read_data_file(args):
+            print('taken from module: {}'.format(module_name))
+        except Exception as e:
+            print('failed importing custom scoring function, error: {}'.format(str(e)))
+            raise
```
Need raise ValueError Here

kuratsak replied:
I guess the exception string itself is enough at this point
Done, I didn't want to lose the traceback inside. is there a trick for this?

weixuanfu2016 commented on tpot/base.py:
```diff
@@ -702,7 +719,29 @@ def set_params(self, **params):
+    def _save_periodic_pipeline(self):
+        if self.periodic_save_path is not None:
+            try:
+                write = self._pbar.write if not self._pbar.disable else print
```
I do not think `print` is needed here when verbosity < 2.

rhiever commented on tpot/base.py:
```diff
@@ -79,12 +80,15 @@ def handler(dwCtrlType, hook_sigint=_thread.interrupt_main):
                  max_time_mins=None, max_eval_time_mins=5,
                  random_state=None, config_dict=None, warm_start=False,
-                 verbosity=0, disable_update_check=False):
+                 verbosity=0, disable_update_check=False, periodic_save_path=None):
```
Please move `periodic_save_path` above `verbosity` and `disable_update_check`.

kuratsak replied:
Done :+1:

rhiever commented on tpot/export_utils.py:
```diff
@@ -88,6 +90,11 @@ def export_pipeline(exported_pipeline, operators, pset):
+    if pscore is not None:
+        pipeline_text += '\n# Score on the training set was:{}\n'.format(pscore)
+    else:
+        pipeline_text += '\n'
```
Please refactor to:

```Python
    if pscore is not None:
         pipeline_text += '\n# Score on the training set was: {}'.format(pscore)
    pipeline_text += '\n'
```

Please also rename `pscore` to `pipeline_score` for a more descriptive variable name.

kuratsak replied:
Done :+1:

rhiever commented on tpot/base.py:
```diff
@@ -702,7 +719,33 @@ def set_params(self, **params):
+    def _save_periodic_pipeline(self):
+        if self.periodic_save_path is not None:
+            try:
+                write = self._pbar.write if not self._pbar.disable else print
```
Please give a more descriptive variable name for `write`.

kuratsak replied:
Done :+1:

rhiever commented on tpot/base.py:
```diff
@@ -702,7 +719,33 @@ def set_params(self, **params):
+                if self.verbosity >= 2:
+                    write('failed saving periodic pipeline, exception:\n{}'.format(str(e)[:250]))
+
+    def export(self, output_file_name, skip_if_repeated=False):
```
Please document `skip_if_repeated` in the docstrings.

kuratsak replied:
Done

rhiever commented on tpot/base.py:
```diff
@@ -718,8 +761,16 @@ def export(self, output_file_name):
+
+        #dont export a pipeline you just had
+        if skip_if_repeated and (self._exported_pipeline_text == to_write):
+            return False
```
Please document the meaning of the `True`/`False` return values for the `export` function in the docstrings.

kuratsak replied:
Done

rhiever commented on tpot/gp_deap.py:
```diff
@@ -164,6 +165,9 @@ def eaMuPlusLambda(population, toolbox, mu, lambda_, cxpb, mutpb, ngen, pbar,
     for gen in range(1, ngen + 1):
+        # after each population save a periodic pipeline
+        if periodic_pipeline_saver is not None:
+            periodic_pipeline_saver()
```
Noting that this addition is related to #79.

rhiever commented on tpot/gp_deap.py:
```diff
@@ -103,7 +103,8 @@ def varOr(population, toolbox, lambda_, cxpb, mutpb):
 def eaMuPlusLambda(population, toolbox, mu, lambda_, cxpb, mutpb, ngen, pbar,
-                   stats=None, halloffame=None, verbose=0, max_time_mins=None):
+                   stats=None, halloffame=None, verbose=0, max_time_mins=None,
+                   periodic_pipeline_saver=None):
```
Please refactor `periodic_pipeline_saver` to `periodic_pipeline_saving_fn` or something more descriptive. Please also document it in the docstrings.

kuratsak replied:
Great idea,
in fact in eaMuPlusLambda context this is a general parameter "per generation function"
renaming to per_generation_hook_func to be more explicit in its role.

This can later be used for more extended logic too.

kuratsak replied:
actually hook seems a bit confusing to some people.
ill just call it per_generation_function

rhiever commented on tpot/base.py:
```diff
@@ -702,7 +719,33 @@ def set_params(self, **params):
         return self
 
-    def export(self, output_file_name):
+    def save_pipeline_if_period(self):
```
`save_pipeline_if_period` should also be a"private" function.

kuratsak replied:
Done

rhiever commented on tpot/base.py:
```diff
@@ -598,6 +614,7 @@ def _update_top_pipeline(self):
                 if pipeline_scores.wvalues[1] > top_score:
                     self._optimized_pipeline = pipeline
                     top_score = pipeline_scores.wvalues[1]
```
`top_score` variable isn't needed any more if the same value is being stored in `self._optimized_pipeline_score`.

kuratsak replied:
removed top_score variable

rhiever commented on tpot/base.py:
```diff
@@ -185,6 +189,13 @@ def __init__(self, generations=100, population_size=100, offspring_size=None,
             A setting of 2 or higher will add a progress bar during the optimization procedure.
         disable_update_check: bool, optional (default: False)
             Flag indicating whether the TPOT version checker should be disabled.
+        periodic_save_path: path string (default: None)
```
Please refactor the `periodic_save_path` name to `periodic_checkpoint_path`. I believe that's a more descriptive term for what it's doing.

kuratsak replied:
changed to periodic_checkpoint_folder to be explicit in its role. Thanks!

rhiever commented on tpot/base.py:
```diff
@@ -79,12 +80,15 @@ def handler(dwCtrlType, hook_sigint=_thread.interrupt_main):
     """Automatically creates and optimizes machine learning pipelines using GP."""
 
+    # dont save periodic pipelines more often than this
+    OUTPUT_BEST_PIPELINE_PERIOD_SECONDS = 30
```
Should this be a TPOT parameter? Thoughts @weixuanfu2016 and @teaearlgraycold? I suppose I'm fine with a hard-coded 30 seconds, but I can see a user down the line wanting to tweak this value.

kuratsak replied:
it's initial target is not to be user configurable - but to prevent too many calls to trying to export. in any case the limit is per generation.

if we one day allow the user to save pipeline after each single evaluation for example, then we might want this (or even might NOT want this).

In any case IMHO it seems a bit early to let this be a parameter.

rhiever commented on tests.py:
```diff
@@ -444,6 +455,36 @@ def test_random_ind_2():
+"""
+
+    assert_equal(expected_code, export_pipeline(pipeline, tpot_obj.operators, tpot_obj._pset, pscore=0.929813743))
+
```
I think these tests look good, but they need to be rebased on the most recent merge. @teaearlgraycold reorganized `tests.py` into a `tests` directory. Less clutter.

kuratsak replied:
Done :+1:

rhiever commented:
Overall, I'm 👍 on these changes, pending the rebase and requested changes.

Beyond the several requests for adding to the docstrings, please also update the [API documentation](http://rhiever.github.io/tpot/api/) and other documentation where relevant.

Any ideas why the coverage decreased by 1.4%?

rhiever commented on tpot/driver.py:
```diff
@@ -404,6 +429,28 @@ def _read_data_file(args):
+            print('failed importing custom scoring function, error: {}'.format(str(e)))
+            raise ValueError(e)
+    
+    return scoring_func
```
Will this function work if I pass a sklearn metric, e.g., `sklearn.metrics.auc`?

kuratsak replied:
added your function in the test, works out of the box :+1:

kuratsak commented:
Yea, I added some ifs on verbosity that another comment requested.
Rebase? I based it on the development branch.
Is that not the most updated one?

On Thu, 15 Jun 2017, 21:31 Randy Olson, <notifications@github.com> wrote:

> Overall, I'm 👍 on these changes, pending the rebase and requested
> changes.
>
> Beyond the several requests for adding to the docstrings, please also
> update the API documentation <http://rhiever.github.io/tpot/api/> and
> other documentation where relevant.
>
> Any ideas why the coverage decreased by 1.4%?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rhiever/tpot/pull/498#issuecomment-308821985>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMzW7HLeMFFWHlseEVgDAdoFFDZVTPfnks5sEXHcgaJpZM4N7Ikm>
> .
>

rhiever commented:
We merged some PRs today that re-organized the unit tests. I think it's just a matter of moving the unit tests into the appropriate files in the `tests` directory.

kuratsak commented:
Is there a branch? I'll merge on Sunday, it's my final work day so last
chance to finish this PR.
I will try to address all the comments then..

On Thu, 15 Jun 2017, 23:14 Randy Olson, <notifications@github.com> wrote:

> We merged some PRs today that re-organized the unit tests. I think it's
> just a matter of moving the unit tests into the appropriate files in the
> tests directory.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/rhiever/tpot/pull/498#issuecomment-308852841>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMzW7LZw7zagiCrviFUdk7pOfLhYSqz0ks5sEZAFgaJpZM4N7Ikm>
> .
>

rhiever commented:
It's merged into the dev branch now.

kuratsak commented:
Done, fixed all comments and remerged development into my branch.
Sadly this is my last day at work, so next time I can fix anything would be
in about a month (mid-july).

I fixed all your comments as you asked, I'm hoping we can merge this now.

On Thu, Jun 15, 2017 at 11:56 PM, Randy Olson <notifications@github.com>
wrote:

> It's merged into the dev branch now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rhiever/tpot/pull/498#issuecomment-308863490>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMzW7H7AyKmJ1yn7yBYCfWVzy_NHjUBCks5sEZqNgaJpZM4N7Ikm>
> .
>



-- 
Cheers,
Dani K.

kuratsak commented:
Can someone please check and rerun travis build? it seems the build failed because of external sources of trouble..

kuratsak commented:
Yey!
Even on the coverage front I look good!

On Sun, Jun 18, 2017 at 12:44 PM, Coveralls <notifications@github.com>
wrote:

> [image: Coverage Status] <https://:/builds/12019579>
>
> Coverage increased (+6.5%) to 94.81% when pulling *81045ab
> <https://github.com/rhiever/tpot/commit/81045ab98cbf366307181c158fcece3a0304936b>
> on kuratsak:tpot_features* into *2b0c29c
> <https://github.com/rhiever/tpot/commit/2b0c29cc1020602f4a89bfa50f0f83386a6db133>
> on rhiever:development*.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rhiever/tpot/pull/498#issuecomment-309267115>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMzW7PQm8VDoocIX-JW78M1CLgZkGqtyks5sFPF6gaJpZM4N7Ikm>
> .
>



-- 
Cheers,
Dani K.

kuratsak commented:
I am being taken away from my working computer, they are saying im fired and i cant touch it anymore :1st_place_medal:

rhiever commented:
Thanks @kuratsak. Will look through this soon.

rhiever commented:
LGTM. Thanks @kuratsak!
